components/datetime/src/format/time_zone.rs

```diff
@@ -37,11 +37,79 @@ where
     T: TimeZoneInput,
 {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        write_pattern(self.time_zone_format, self.time_zone, f).map_err(|_| core::fmt::Error)
+        write_zone(self.time_zone_format, self.time_zone, f).map_err(|_| core::fmt::Error)
     }
 }
 
-// TODO(#622) Make this public once TimeZoneFormat is public.
+pub(crate) fn write_zone<T, W>(
+    time_zone_format: &TimeZoneFormat,
+    time_zone: &T,
+    w: &mut W,
+) -> Result<(), Error>
+where
+    T: TimeZoneInput,
+    W: fmt::Write + ?Sized,
+{
+    match (&time_zone_format.patterns, &time_zone_format.config) {
+        (Some(_), None) => write_pattern(time_zone_format, time_zone, w),
+        (None, Some(_)) => write_config(time_zone_format, time_zone, w),
+        _ => todo!("Invalid state (both populated, or none populated, shouldn't be possible)"),
+    }
+}
+
+pub(crate) fn write_config<T, W>(
+    time_zone_format: &TimeZoneFormat,
+    time_zone: &T,
+    w: &mut W,
+) -> Result<(), Error>
+where
+    T: TimeZoneInput,
+    W: fmt::Write + ?Sized,
+{
+    match time_zone_format.config.as_ref().unwrap() {
+        TimeZoneFormatConfig::GenericNonLocationLong => {
+            time_zone_format
+                .long_generic_non_location_format(w, time_zone)
+                .or_else(|_| time_zone_format.generic_location_format(w, time_zone))
+                .or_else(|_| time_zone_format.localized_gmt_format(w, time_zone))?;
+        }
+        TimeZoneFormatConfig::GenericNonLocationShort => {
+            time_zone_format
+                .short_generic_non_location_format(w, time_zone)
+                .or_else(|_| time_zone_format.generic_location_format(w, time_zone))
+                .or_else(|_| time_zone_format.localized_gmt_format(w, time_zone))?;
+        }
+        TimeZoneFormatConfig::GenericLocation => {
+            time_zone_format
+                .generic_location_format(w, time_zone)
+                .or_else(|_| time_zone_format.localized_gmt_format(w, time_zone))?;
+        }
+        TimeZoneFormatConfig::SpecificNonLocationLong => {
+            time_zone_format
+                .long_specific_non_location_format(w, time_zone)
+                .or_else(|_| time_zone_format.localized_gmt_format(w, time_zone))?;
+        }
+        TimeZoneFormatConfig::SpecificNonLocationShort => {
+            time_zone_format
+                .short_specific_non_location_format(w, time_zone)
+                .or_else(|_| time_zone_format.localized_gmt_format(w, time_zone))?;
+        }
+        TimeZoneFormatConfig::LocalizedGMT => {
+            time_zone_format.localized_gmt_format(w, time_zone)?;
+        }
+        TimeZoneFormatConfig::Iso8601(iso_format, iso_minutes, iso_seconds) => {
+            time_zone_format.iso8601_format(
+                w,
+                time_zone,
+                iso_format.clone(),
+                iso_minutes.clone(),
+                iso_seconds.clone(),
```

Comment:
> suggestion

Once we remove the `as_ref()` in the comment above, we can remove all of these `.clone()` calls, since everything will be trivially copied.